### PR TITLE
Accept multiple SSH keys in single LDAP SSHPublicKey attribute

### LINKED ISSUE #13984: Sanitize SSH pubkeys (and Support multiple) when sync from LDAP
In our setup we allow multiple SSH keys to be entered in the LDAP database, for each user.
When Gitea synchonizes those keys it does so by creating a wrong configuration, including newlines (which separate the SSH keys) in the authorized_key, basically allowing direct ssh access.

### PR DESCRIPTION
**Pull request by zeripath**
Fix #13984

Signed-off-by: Andrew Thornton <art27@cantab.net>

### CODE REVIEW

File: models/user.go [hunk 1]
```diff
@@ -1588,9 +1588,23 @@ func deleteKeysMarkedForDeletion(keys []string) (bool, error) {
...
+			out, comment, _, keys, err = ssh.ParseAuthorizedKey(keys)
+			if err != nil {
+				break loop
+			}
+			found = true
+			marshalled := out.Marshal()
+			sshKeyName := fmt.Sprintf("%s-%s", s.Name, marshalled[0:40])
+
+			marshalled = append(marshalled, []byte(" "+comment)...)
 			if _, err := AddPublicKey(usr.ID, sshKeyName, sshKey, s.ID); err != nil {
```
Comment (zeripath) [inline models/user.go]:
```suggestion
			if _, err := AddPublicKey(usr.ID, sshKeyName, string(marshalled), s.ID); err != nil {
```

**Comment by strk** [comment]
Thanks, I'll try to test this out, but I'm blocked by #13993 at the moment.
I see there's an `integrations/auth_ldap_test.go` file which could be used to add automated testing for this fix.
Would be nice to test some malicious ssh pub keys, like containing pipes and comments and what not (injection attempts)

**Comment by lafriks** [comment]
tests fail now

**Comment by zeripath** [comment]
UGh this is quite a bit more fiddly than I initially expected.

**Comment by zeripath** [comment]
Turns out that the last problem was that the marshall authorized key added a newline - which was the cause of the failing test!

**Comment by strk** [comment]
Great that you handled to fix the existing tests, but would you please consider adding a _new_ test for having multiple keys and malicious code in single elements coming from LDAP ? That was the issue reported in #13984

**Comment by strk** [comment]
Great work ! LGTM

**Review by lunny — APPROVED** [review]

**Review by lafriks — APPROVED** [review]

**Pull request merged**